File: fragment/ListFragment.ts
import { HTMLText, Text } from "pixi.js";
import Fragment, { decode, FragmentType, register } from "./Fragment";

//@ts-ignore
import * as wasm from "../WasmEndec-1.0-SNAPSHOT/js/endec";



const LIST = register("trickster:list", 0xffffff, (object: any) => {
    if (object instanceof wasm.ListFragment) {
        const fragments: Array<Fragment | null> = object.fragments.map(decode);

        if (fragments.every((x) => x instanceof Fragment)) {
            return new ListFragment(fragments);
        }
    }
    return null;
});

export default class ListFragment extends Fragment {
    fragments: Array<Fragment>;

    constructor(fragments: Array<Fragment>) {
        super();

        this.fragments = fragments;
    }

    override asFormattedText(): HTMLText {
        let result = "[";

        for (let i = 0; i < this.fragments.length; i++) {
            const fragment = this.fragments[i];
            if (i !== 0) {
                result += ", ";
            }
            result += fragment.asFormattedText().text;
        }

        result += "]";

        return new HTMLText({
            text: `<span style="color: #${this.type().color.toString(16)}">${result}</span>`,
        });
    }

    override toString(): string {
        let result = "[";

        for (let i = 0; i < this.fragments.length; i++) {
            const fragment = this.fragments[i];
            if (i !== 0) {
                result += ", ";
            }
            result += fragment.asFormattedText().text;
        }

        result += "]";

        return result;
    }

    override type(): FragmentType {
        return LIST;
    }
}

File: src/jsMain/kotlin/me/maplesyrum/tricksterstudio/spell/fragment/ListFragment.kt
package me.maplesyrum.tricksterstudio.spell.fragment


import io.kvision.utils.obj
import me.maplesyrum.tricksterstudio.external.pixi.HTMLText
import tree.maple.kendec.StructEndec
import tree.maple.kendec.impl.StructEndecBuilder


class ListFragment(val fragments: List<Fragment>) : Fragment() {
    override fun type(): FragmentType<*> {
        return FragmentType.LIST
    }

    override fun asFormattedText(): HTMLText {
        val result = buildString {
            append("[")
            for ((i, fragment) in fragments.withIndex()) {
                if (i != 0) append(", ")
                append(fragment.asFormattedText().text)
            }
            append("]")
        }
        return HTMLText(
           obj {
               text = "<span style=\"color: #${type().color.map(::toInt).orElse(0xaaaaaa).toString(16)}\">$result</span>"
           }
        )
    }

    override fun toString(): String {
        return buildString {
            append("[")
            for ((i, fragment) in fragments.withIndex()) {
                if (i != 0) append(", ")
                append(fragment.asFormattedText().text)
            }
            append("]")
        }
    }

    companion object {
        val ENDEC: StructEndec<ListFragment> = StructEndecBuilder.of(
            Fragment.ENDEC.listOf().fieldOf("fragments", ListFragment::fragments)
        ) { fragments: List<Fragment> -> ListFragment(fragments) }
    }
}
